new-log-viewer/src/services/decoders/JsonlDecoder/utils.ts

```diff
@@ -0,0 +1,87 @@
+import dayjs, {Dayjs} from "dayjs";
+
+import {
+    JsonObject,
+    JsonValue,
+} from "../../../typings/js";
+import {
+    INVALID_TIMESTAMP_VALUE,
+    LOG_LEVEL,
+} from "../../../typings/logs";
+
+
+/**
+ * Determines whether the given value is a `JsonObject` and applies a TypeScript narrowing
+ * conversion if so.
+ *
+ * @param value
+ * @return A TypeScript type predicate indicating whether `value` is a `JsonObject`.
+ */
+const isJsonObject = (value: JsonValue): value is JsonObject => {
+    return "object" === typeof value && null !== value;
+};
+
+/**
+ * Converts a field into a log level if possible.
+ *
+ * @param field
+ * @return The log level or `LOG_LEVEL.NONE` if the field couldn't be converted.
+ */
+const convertToLogLevelValue = (field: JsonValue | undefined): LOG_LEVEL => {
+    let logLevelValue = LOG_LEVEL.NONE;
+
+    if ("undefined" === typeof field) {
+        return logLevelValue;
+    }
+
+    // Json stringify covers edge case where the field is an object with more than one key, e.g.,
+    // `field = { "name": "INFO", "value": 20 }`.
+    const logLevelName = "object" === typeof field ?
+        JSON.stringify(field) :
```

Comment:
@junhaoliao, if it's an object, shouldn't we just return early? A stringified JSON object won't match any of our hard-coded log levels, right?

Reply:
If we are to support stringified JSON objects, the code below should be written as 
```
    const uppercaseLogLevelName = logLevelName.toUpperCase();
    logLevelValue = LOG_LEVEL_NAMES.splice(1).find((name) => 
        uppercaseLogLevelName.includes(name)
    ) ?? LOG_LEVEL.NONE;
```

I vaguely remember seeing log levels being stored nested as an object in some logs though I can't find the logs now. If you believe the code has too much overhead introduced, let's do early returns, and we can come back to this for a more properly implementation when we see such requirements.

Reply:
Gotcha. Iirc, the MongoDB log dataset has the level nested in key. That said, I'm not sure we should bother stringifying the field if it's an object since:

1. It introduces overhead.
2. It's a heuristic to search for the log level string in the stringified object.
3. It hides the issue from the user which is really that they should be specifying the nested key as the level field.

What do you think?

Reply:
Sound good. Let's do early return here and defer the requirement of such nested objects.

Just curious, how does clp-s handle timestamps from nested objects? e.g., how should the user specify a timestamp key if the log looks like this:
```json
{
    "t":
    {
        "timestamp.value": 123
    }
}
```

Reply:
Theoretically, the user would escape any periods that are part of a key. So in the example: `t.timestamp\.value`. That said, clp-s doesn't support querying keys with escaped characters yet.